File: native/ios/FireballNative/Core/LastFmClient.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif
#if canImport(CryptoKit)
import CryptoKit
#endif

struct LastFmClient {
    private static let base = URL(string: "https://ws.audioscrobbler.com/2.0/")!

    func validateApiKey(_ apiKey: String) async -> Bool {
        guard !apiKey.isEmpty,
              let url = URL(string: "\(Self.base)?method=chart.gettopartists&limit=1&api_key=\(apiKey.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? apiKey)&format=json")
        else { return false }
        do {
            let (data, _) = try await URLSession.shared.data(from: url)
            guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else { return false }
            return json["error"] == nil
        } catch {
            return false
        }
    }

    func mobileSession(apiKey: String, apiSecret: String, username: String, password: String) async -> String? {
        guard !apiKey.isEmpty, !apiSecret.isEmpty, !username.isEmpty, !password.isEmpty else { return nil }
        var params: [String: String] = [
            "method": "auth.getMobileSession",
            "username": username,
            "password": md5(password),
            "api_key": apiKey,
        ]
        params["api_sig"] = sign(params, secret: apiSecret)
        guard let url = signedURL(params: params) else { return nil }
        do {
            let (data, _) = try await URLSession.shared.data(from: url)
            guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let session = json["session"] as? [String: Any],
                  let key = session["key"] as? String
            else { return nil }
            return key
        } catch {
            return nil
        }
    }

    func updateNowPlaying(apiKey: String, apiSecret: String, sessionKey: String, title: String, artist: String, album: String?) async -> Bool {
        var params: [String: String] = [
            "method": "track.updateNowPlaying",
            "api_key": apiKey,
            "sk": sessionKey,
            "artist": artist,
            "track": title,
        ]
        if let album, !album.isEmpty { params["album"] = album }
        return await postSigned(params: params, apiSecret: apiSecret)
    }

    func scrobble(apiKey: String, apiSecret: String, sessionKey: String, title: String, artist: String, listenedAt: Int64, album: String?) async -> Bool {
        var params: [String: String] = [
            "method": "track.scrobble",
            "api_key": apiKey,
            "sk": sessionKey,
            "artist": artist,
            "track": title,
            "timestamp": String(listenedAt),
        ]
        if let album, !album.isEmpty { params["album"] = album }
        return await postSigned(params: params, apiSecret: apiSecret)
    }

    private func postSigned(params: [String: String], apiSecret: String) async -> Bool {
        var signed = params
        signed["api_sig"] = sign(params, secret: apiSecret)
        guard let url = signedURL(params: signed) else { return false }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        do {
            let (data, _) = try await URLSession.shared.data(for: request)
            guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else { return false }
            return json["error"] == nil
        } catch {
            return false
        }
    }

    private func signedURL(params: [String: String]) -> URL? {
        var components = URLComponents(url: Self.base, resolvingAgainstBaseURL: false)
        components?.queryItems = params.map { URLQueryItem(name: $0.key, value: $0.value) } + [URLQueryItem(name: "format", value: "json")]
        return components?.url
    }

    private func sign(_ params: [String: String], secret: String) -> String {
        let concat = params.keys.sorted().map { $0 + (params[$0] ?? "") }.joined() + secret
        return md5(concat)
    }

    private func md5(_ input: String) -> String {
        #if canImport(CryptoKit)
        return Insecure.MD5.hash(data: Data(input.utf8)).map { String(format: "%02hhx", $0) }.joined()
        #else
        // Linux CI compile path; Last.fm is exercised on Apple devices in production builds.
        return ""
        #endif
    }
}

File: native/android/core-data/src/main/kotlin/com/fireball/nativeapp/core/data/integrations/LastFmClient.kt
package com.fireball.nativeapp.core.data.integrations

import io.ktor.client.HttpClient
import io.ktor.client.request.get
import io.ktor.client.request.parameter
import io.ktor.client.request.post
import io.ktor.client.request.url
import io.ktor.client.statement.bodyAsText
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.security.MessageDigest
import java.util.SortedMap

/**
 * Last.fm Audioscrobbler API. Scrobbling requires [getMobileSession] (API key + shared secret + credentials).
 */
class LastFmClient(private val httpClient: HttpClient) {
    private val json = Json { ignoreUnknownKeys = true }

    suspend fun validateApiKey(apiKey: String): Boolean {
        if (apiKey.isBlank()) return false
        return runCatching {
            val body = httpClient.get {
                url(BASE)
                parameter("method", "chart.gettopartists")
                parameter("api_key", apiKey)
                parameter("limit", "1")
                parameter("format", "json")
            }.bodyAsText()
            val root = json.parseToJsonElement(body).jsonObject
            !root.containsKey("error")
        }.getOrDefault(false)
    }

    suspend fun getMobileSession(
        apiKey: String,
        apiSecret: String,
        username: String,
        password: String,
    ): String? {
        if (apiKey.isBlank() || apiSecret.isBlank() || username.isBlank() || password.isBlank()) return null
        val params = sortedMapOf(
            "method" to "auth.getMobileSession",
            "username" to username,
            "password" to md5(password),
            "api_key" to apiKey,
        )
        val apiSig = buildSignature(params, apiSecret)
        return runCatching {
            val body = httpClient.get {
                url(BASE)
                params.forEach { entry -> parameter(entry.key, entry.value) }
                parameter("api_sig", apiSig)
                parameter("format", "json")
            }.bodyAsText()
            val root = json.parseToJsonElement(body).jsonObject
            root["session"]?.jsonObject?.get("key")?.jsonPrimitive?.content
        }.getOrNull()
    }

    suspend fun updateNowPlaying(
        apiKey: String,
        apiSecret: String,
        sessionKey: String,
        title: String,
        artist: String,
        album: String? = null,
    ): Boolean {
        if (apiKey.isBlank() || apiSecret.isBlank() || sessionKey.isBlank()) return false
        val params = sortedMapOf(
            "method" to "track.updateNowPlaying",
            "api_key" to apiKey,
            "sk" to sessionKey,
            "artist" to artist,
            "track" to title,
        )
        if (!album.isNullOrBlank()) params["album"] = album
        return postSigned(params, apiSecret)
    }

    suspend fun scrobble(
        apiKey: String,
        apiSecret: String,
        sessionKey: String,
        title: String,
        artist: String,
        timestampEpoch: Long,
        album: String? = null,
    ): Boolean {
        if (apiKey.isBlank() || apiSecret.isBlank() || sessionKey.isBlank()) return false
        val params = sortedMapOf(
            "method" to "track.scrobble",
            "api_key" to apiKey,
            "sk" to sessionKey,
            "artist" to artist,
            "track" to title,
            "timestamp" to timestampEpoch.toString(),
        )
        if (!album.isNullOrBlank()) params["album"] = album
        return postSigned(params, apiSecret)
    }

    private suspend fun postSigned(params: Map<String, String>, apiSecret: String): Boolean {
        val apiSig = buildSignature(params, apiSecret)
        return runCatching {
            val body = httpClient.post {
                url(BASE)
                params.forEach { entry -> parameter(entry.key, entry.value) }
                parameter("api_sig", apiSig)
                parameter("format", "json")
            }.bodyAsText()
            val root = json.parseToJsonElement(body).jsonObject
            !root.containsKey("error")
        }.getOrDefault(false)
    }

    private fun buildSignature(params: Map<String, String>, secret: String): String {
        val concat = params.toSortedMap().entries.joinToString("") { it.key + it.value } + secret
        return md5(concat)
    }

    private fun md5(input: String): String {
        val digest = MessageDigest.getInstance("MD5").digest(input.toByteArray(Charsets.UTF_8))
        return digest.joinToString("") { "%02x".format(it) }
    }

    private companion object {
        const val BASE = "https://ws.audioscrobbler.com/2.0/"
    }
}
